core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -90,6 +93,33 @@ public static ExplicitReturnTypeInference explicit(SqlTypeName typeName,
    */
   public static final SqlReturnTypeInference ARG0 =
       new OrdinalReturnTypeInference(0);
+
+  /**
+   * Type-inference strategy whereby the result type of a table function call is a ROW,
+   * which is combined from the operand #0(TABLE parameter)'s schema and two
+   * additional fields:
+   *
+   * <ol>
+   *  <li>window_start. TIMESTAMP type to indicate a window's start.</li>
+   *  <li>window_end. TIMESTAMP type to indicate a window's end.</li>
+   * </ol>
+   */
+  public static final SqlReturnTypeInference ARG0_TABLE_FUNCTION_WINDOWING =
+      opBinding -> {
+        RelDataType inputRowType = opBinding.getOperandType(0);
+        List<RelDataTypeField> newFields = new ArrayList<>(inputRowType.getFieldList());
+        RelDataType timestampType = opBinding.getTypeFactory().createSqlType(SqlTypeName.TIMESTAMP);
+
+        RelDataTypeField windowStartField =
+            new RelDataTypeFieldImpl("window_start", newFields.size(), timestampType);
+        newFields.add(windowStartField);
+        RelDataTypeField windowEndField =
+            new RelDataTypeFieldImpl("window_end", newFields.size(), timestampType);
```

Comment:
Move this return type check logic into `SqlTableValuedFunctionWindowingOperator` directly.

Reply:
done